icicle/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,663 @@
+    task->set_phase1_addition_with_affine(bkt, base, task_bkt_idx);
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase1_wait_for_completion()
+{
+  EcAddTask<A, P>* task = manager.get_completed_task();
+  while (task != nullptr) {
+    // Check for collision in the destination bucket, and chain and addition / store result accordingly.
+    if (m_bkts_occupancy[task->m_return_idx]) {
+      m_bkts_occupancy[task->m_return_idx] = false;
+      task->set_phase1_collision_task(m_buckets[task->m_return_idx]);
+    } else {
+      m_buckets[task->m_return_idx] = task->m_p1;
+      m_bkts_occupancy[task->m_return_idx] = true;
+      task->set_idle();
+    }
+    task = manager.get_completed_task();
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase2_bm_sum(std::vector<BmSumSegment>& segments)
+{
+  phase2_setup(segments);
+  if (m_segment_size > 1) {
+    // Send first additions - line additions.
+    for (int i = 0; i < m_num_bms * m_num_bm_segments; i++) {
+      EcAddTask<A, P>* task = manager.get_idle_task();
+      BmSumSegment& curr_segment = segments[i]; // For readability
+
+      int bkt_idx = curr_segment.m_segment_mem_start + curr_segment.m_idx_in_segment;
+      P bucket = m_bkts_occupancy[bkt_idx] ? m_buckets[bkt_idx] : P::zero();
+      task->set_phase2_addition_by_value(curr_segment.line_sum, bucket, i);
+    }
+
+    // Loop until all line/tri sums are done.
+    int done_segments = 0;
+    while (done_segments < m_num_bms * m_num_bm_segments) {
+      EcAddTask<A, P>* task = manager.get_completed_task();
+      BmSumSegment& curr_segment = segments[task->m_return_idx]; // For readability
+
+      if (task->m_is_line) {
+        curr_segment.line_sum = task->m_p1;
+      } else {
+        curr_segment.triangle_sum = task->m_p1;
+      }
+      curr_segment.m_nof_received_sums++;
+
+      // Check if this was the last addition in the segment
+      if (curr_segment.m_idx_in_segment < 0) {
+        done_segments++;
+        task->set_idle();
+        continue;
+      }
+      // Otherwise check if it is possible to assign new additions:
+      // Triangle sum is dependent on the 2 previous sums (line and triangle) - so check if 2 sums were received.
+      if (curr_segment.m_nof_received_sums == 2) {
+        curr_segment.m_nof_received_sums = 0;
+        task->set_phase2_triangle_addition(curr_segment.line_sum, &(curr_segment.triangle_sum));
+        curr_segment.m_idx_in_segment--;
+
+        // Line sum (if not the last one in the segment)
+        if (curr_segment.m_idx_in_segment >= 0) {
+          int bkt_idx = curr_segment.m_segment_mem_start + curr_segment.m_idx_in_segment;
+          if (m_bkts_occupancy[bkt_idx]) {
+            int return_idx = task->m_return_idx;
+            // Due to the choice of num segments being less than half of total tasks there ought to be an idle task for
+            // the line sum
+            task = manager.get_idle_task();
```

Comment:
as we discussed, use get idle for both tasks or assign thread per segment in advance